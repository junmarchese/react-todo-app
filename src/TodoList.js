import React, { useState } from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";

function TodoList() {
    const [todos, setTodos] = useState([]);

    const addTodo = (newTodo) => {
        setTodos((todos) => [...todos, newTodo]);
    };

    const removeTodo = (id) => {
        setTodos((todos) => todos.filter((todo) => todo.id !== id));
    };

    return (
        <div>
            <NewTodoForm addTodo={addTodo} />
            <ul>
                {todos.map((todo) => (
                    <li key={todo.id}>
                        <Todo id={todo.id} task={todo.task} removeTodo={removeTodo} />
                    </li>
                ))}
            </ul>
        </div>
    );
}

export default TodoList;
