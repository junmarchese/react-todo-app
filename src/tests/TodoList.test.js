import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TodoList from "../TodoList";

it("renders without crashing", () => {
    render(<TodoList />);
});

it("matches snapshot", () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

describe("TodoList functionality", () => {
    it("can add a new todo", async () => {
        render(<TodoList />);

        const input = screen.getByLabelText("New Todo:");
        const addButton = screen.getByText("Add Todo");

        await userEvent.type(input, "Test Todo");
        await userEvent.click(addButton);

        expect(await screen.findByText("Test Todo")).toBeInTheDocument();
    });

    it("can remove a todo", async () => {
        render(<TodoList />);

        const input = screen.getByLabelText("New Todo:");
        const addButton = screen.getByText("Add Todo");

        await userEvent.type(input, "Test Todo");
        await userEvent.click(addButton);

        const todoText = await screen.findByText("Test Todo");
        expect(todoText).toBeInTheDocument();

        const deleteButton = await screen.findByText("X");
        await userEvent.click(deleteButton);

        expect(todoText).not.toBeInTheDocument();
    });

    it("does not add empty todos", () => {
        render(<TodoList />);

        const addButton = screen.getByText("Add Todo");
        userEvent.click(addButton);

        const todoItems = screen.queryAllByRole("listitem");
        expect(todoItems.length).toBe(0);
    });
});
