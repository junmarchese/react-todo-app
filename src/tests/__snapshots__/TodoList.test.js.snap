// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="task"
      >
        <strong>
          New Todo:
        </strong>
      </label>
      <input
        id="task"
        name="task"
        value=""
      />
      <button>
        Add Todo
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
